src/Microsoft.ApplicationInsights/DataContracts/TelemetryContext.cs

```diff
@@ -96,7 +96,7 @@ public DeviceContext Device
         /// </summary>
         public CloudContext Cloud
         {
-            get { return LazyInitializer.EnsureInitialized(ref this.cloud, () => new CloudContext()); }
+            get { return this.cloud ?? (this.cloud = new CloudContext()); }
```

Comment:
It is JIT_WriteBarrier, Interlocked.CompareExchange and related (not lambda function construction). My concern here as well that this code is "pseudo" race condition proof. For example, later on it is used in a following way:
```           var cloudContext = context.Cloud;
            if (string.IsNullOrEmpty(cloudContext.RoleName))
            {
                cloudContext.RoleName = this.roleName;
            }
```
which is not properly synchronized anyway.
But in either case - I've reviewed the perf impact of this, and decided that we can take it as part of another PR, where we would need to carefully review usage of Interlocked.Exchange.

Regarding the lambda creator vs omitting it - it needs more investigation. In particular: 1) it might use the `Activator.CreateInstance()` which is slower 2) what is the current code being actually generated - are there any allocations (according to decompiled code the Func<CloudContext> is actually cached as a class member). 3) And of course - actual perf measurments for each case. We will do the investigation if needed in a separate effort.